docs/proposals/unified-backend-architecture/tech-spec.md

```diff
@@ -0,0 +1,270 @@
+
+## Multi-Tenant Architecture
+
+Mobile clients will communicate with a clusters of nodes, so the basic requirement is an ability of one node to work
+with particular number of clients simultaneously. For example, if 1 cluster has to serve 10000 mobile clients and
+this cluster consists of 100 nodes, each node has to be able to work with 100 clients simultaneously and reliably.
+Thus, the longest procedure is a new payment (creating new transaction), and even if all these 100 clients will
+sign their transactions and send it for publishing at the same time, the node must be able to handle these
+transactions reliably and as fast as possible.
+
+To achieve this goal the node must implement robust concurrent model. More specifically, all requests to the wallet
+web API endpoints must be handled concurrently. In addition, it is possible to use queues/buffers, it will allow
+to keep sustained transactions submission rate.
+
+In case of correct design decisions, even the slowest endpoints, like `/api/v1/transactions`, can be handled fully
+concurrently (publishing to the network, storing in the `wallet-db`, etc).
+
+_Alfredo said: AFAIK that's possible, `acid-state` supports atomic transactions and all the rest of it. Provided we
+write the right `Update` transactions we should be fine._
+
+## Scalability
+
+Initially clusters of nodes will be hosted by IOHK (on AWS EC2). And we need a mechanism of auto-scaling of the
+cluster: it must be able to grow proportionally to growing number of clients.
+
+**TODO**: To think about auto-scale supervisor. Suppose we have 1 cluster that consists of 100 nodes and each of node
+can reliably serve 100 clients. If supervisor sees that number of clients connected to this cluster is growing (and
+becomes more than 10000), it launches additional nodes. But exact resource and tuning to do to be decided during
+later benchmarking.
+
+_Domen said: We might not base our autoscaling on number of users, but another resource. It's hard to say how it
+would scale before we run some benchmarks. It could be that bottleneck is CPU and you can't correlate that with number
+of subscribers._
+
+**TODO 2**: Launching additional node is relatively expensive action (in terms of time and memory). So it is better
+if one single node will be able to serve more clients than just increase a number of nodes in the cluster.
+We have to understand what the real bottlnecks are. E.g. for relays, the number of clients syncing concurrently
+_is_ the bottleneck, not the number of client just keeping up with the blockchain.
+
+## User Stories
+
+These are typical user stories (workflows), with description of API calls.
+
+### Initial State
+
+After installation of mobile client application user does not see any wallets, and he must create at least one to
+work with a client application.
+
+During the first connection with the node the new session is establishing, please read an explanation below.
+
+### Wallet Creation
```

Comment:
I would rephrase the user stories in the usual style `As a user, I want to do X Y and Z to achieve W`. An example: https://www.romanpichler.com/blog/10-tips-writing-good-user-stories/ (It's quite comprehensive, but I guess you can extract the gist of it). This helps enormously because it helps you reason on the end goal of the user rather than the outcome (in this case the creation of a new wallet).